README
NAME
    Compress::LZF - extremely light-weight Lempel-Ziv-Free compression

SYNOPSIS
       # import compress/decompress functions
       use Compress::LZF;
       # the same as above
       use Compress::LZF ':compress';

       $compressed = compress $uncompressed_data;
       $original_data = decompress $compressed;

       # import sfreeze, sfreeze_cref and sfreeze_c
       use Compress::LZF ':freeze';

       $serialized = sfreeze_c [4,5,6];
       $original_data = sthaw $serialized;

DESCRIPTION
    LZF is an extremely fast (not that much slower than a pure memcpy)
    compression algorithm. It is ideal for applications where you want to
    save *some* space but not at the cost of speed. It is ideal for
    repetitive data as well. The module is self-contained and very small (no
    large library to be pulled in). It is also free, so there should be no
    problems incorporating this module into commercial programs.

    I have no idea wether any patents in any countries apply to this
    algorithm, but at the moment it is believed that it is free from any
    patents.

FUNCTIONS
    $compressed = compress $uncompressed
    $compressed = compress_best $uncompressed
        Try to compress the given string as quickly and as much as possible.
        In the worst case, the string can enlarge by 1 byte, but that should
        be the absolute exception. You can expect a 45% compression ratio on
        large, binary strings.

        The "compress_best" function uses a different algorithm that is
        slower but usually achieves better compression.

    $decompressed = decompress $compressed
        Uncompress the string (compressed by "compress") and return the
        original data. Decompression errors can result in either broken data
        (there is no checksum kept) or a runtime error.

    $serialized = sfreeze $value (simplified freeze)
    $serialized = sfreeze_best $value
        Often there is the need to serialize data into a string. This
        function does that, by using the Storable module. It does the
        following transforms:

          undef (the perl undefined value)
             => a special cookie (undef'ness is being preserved)
          IV, NV, PV (i.e. a _plain_ perl scalar):
             => stays as is when it contains normal text/numbers
             => gets serialized into a string
          RV, undef, other funny objects (magical ones for example):
             => data structure is freeze'd into a string.

        That is, it tries to leave "normal", human-readable data untouched
        but still serializes complex data structures into strings. The idea
        is to keep readability as high as possible, and in cases readability
        can't be helped anyways, it tries to compress the string.

        The "sfreeze" functions will enlarge the original data one byte at
        most and will only load the Storable method when neccessary.

        The "sfreeze_best" function uses a different algorithm that is
        slower but usually achieves better compression.

    $serialized = sfreeze_c $value (sfreeze and compress)
    $serialized = sfreeze_c_best $value
        Similar to "sfreeze", but always tries to "c"ompress the resulting
        string. This still leaves most small objects (most numbers)
        untouched.

        The "sfreeze_c" function uses a different algorithm that is slower
        but usually achieves better compression.

    $serialized = sfreeze_cr $value (sfreeze and compress references)
    $serialized = sfreeze_cr_best $value
        Similar to "sfreeze", but tries to "c"ompress the resulting string
        unless it's a "simple" string. References for example are not
        "simple" and as such are being compressed.

        The "sfreeze_cr_best" function uses a different algorithm that is
        slower but usually achieves better compression.

    $original_data = sthaw $serialized
        Recreate the original object from it's serialized representation.
        This function automatically detects all the different sfreeze
        formats.

    Compress::LZF::set_serializer $package, $freeze, $thaw
        Set the serialize module and functions to use. The default is
        "Storable", "Storable::net_mstore" and "Storable::mretrieve", which
        should be fine for most purposes.

SEE ALSO
    Other Compress::* modules, especially Compress::LZV1 (an older, less
    speedy module that guarentees only 1 byte overhead worst case) and
    Compress::Zlib.

    http://liblzf.plan9.de/

AUTHOR
    This perl extension and the underlying liblzf were written by Marc
    Lehmann <schmorp@schmorp.de> (See also http://liblzf.plan9.de/).

BUGS


Changes
Revision history for Perl extension Compress::LZF

3.7  Sun Aug 25 20:12:58 CEST 2013
	- update liblzf to version 3.7 (speedups, no bugfixes).
        - add _best versions that use lzf_c_best.

3.43 Mon Dec 29 09:46:17 CET 2008
	- port to perls that pointlessly have enabled windows process
          emulation on unix (broguht to my attention by Guillaume Rousse).

3.42 Wed Dec 17 01:35:47 CET 2008
        - grr, require_pv doesn't work either, roll our own.
	- name the correct function serialiser function name
          in error messages.

3.41 Tue Sep 23 01:59:02 CEST 2008
	- it would be nice if the fact that load_module eats
          the module name were documented :(, in any case,
          switching to require_pv fixes all the mysterious
          crashes.
	- be more forgiving instead of crashing when the
          serialiser doesn't return a string.

3.4  Fri Sep  5 17:27:10 CEST 2008
	- update to liblzf-3.4.

3.11 Wed Jan  9 22:43:17 CET 2008
	- add PUTBACK/SPAGAIN around need_storable calls, fixing
          a potential stack change problem when loading the serialiser.

3.1  Fri Nov 30 11:35:47 CET 2007
	- update to liblzf-3.1.

2.0  Tue Nov 13 22:49:57 CET 2007
	- update to liblzf-3.0.

1.9  Fri Nov  2 13:34:21 CET 2007
	- require perl 5.008.
        - updated to liblzf version 2.1.

1.8  Wed Apr 18 14:40:20 CEST 2007
	- properly downgrade all scalars to octet form when
          required.
        - strings potentially containing non-latin1 characters
          get frozen via the serialiser.

1.71 Fri Feb 16 23:09:04 CET 2007
	- fix memory leak when feeding illegal data to decompress.
        - allocate hash dynamically - slower, but having 256k of stack
          is not ubiquitiously available.

1.7  Wed Sep 27 17:52:39 CEST 2006
	- fix buffer overflows in decompress header.

1.65 Fri Jul  7 17:42:12 CEST 2006
	- upgrade to liblzf-1.6.

1.6  Sat Oct  1 03:34:11 CEST 2005
        - now really allow non-Storable serialisers.

1.52 Sat Oct  1 03:23:52 CEST 2005
	- changed mstore to net_mstore default.

1.51 Thu Apr 14 18:59:03 CEST 2005
	- upped compress size limit from 64MB to 2GB (reported
          by Jan Sundberg, who actually ran into the limit :).
        - *blush*, changed Zev-Limpel to Lempel-Ziv... funny that
          nobody spotted thta braino.

1.5   Tue Mar  8 21:04:23 CET 2005
	- upgraded to liblzf-1.5 (faster, slightly worse compression).

1.401 Thu Mar  3 18:00:52 CET 2005
	- upgraded to liblzf-1.401.
	- change of contact address.

1.4   Wed Dec 15 08:09:33 CET 2004
	- upgraded to liblzf-1.4.

1.0   Mon May  3 20:25:07 CEST 2004
	- yet another minor reference copying vs. inc bug.

0.1046 Fri Jan 17 08:14:30 CET 2003
	- added debian files, realized that debians package management
          is quite useless.

0.1045 Thu Sep  5 23:54:11 CEST 2002
	- always make a stringified copy of the argument for
          sfreeze, so the assertion "sfreeze makes a string out
          of it's argument" holds even for integers.
        - "simple" objects now include PVIV and PVNV.

0.1044 Thu Sep  5 06:44:13 CEST 2002
	- MINOR: add set_serializer interface.

0.1043 Sat May 18 23:40:01 CEST 2002
	- bugfix when freezing non-SVt_PV-PV's.
        - thanks to all cpan-testers ;)

0.1042 Fri May 17 22:09:08 CEST 2002
	- removed some debugging code that kept sfreeze from properly

0.1041 Wed Apr 10 05:12:47 CEST 2002
	- sthaw now returns a copy of it's argument instead of the argument
          itself in the pass-through case, as this is the required semantics
          for perl.

0.104 Tue Mar 26 18:48:02 CET 2002
	- fix more logic bugs. never wrote functions with such contorted
          logic :( maybe I should just ignore the idea of saving single bytes
          whereever possible to make this module much cleaner...

0.103 Sun Mar  3 05:39:00 CET 2002
	- try to handle magical objects that are no references
          by creating a referebce to them and passing this ref to
          storable.

0.102 Wed Feb 27 21:50:18 CET 2002
	- handle magic in sfreeze.

0.101 Fri Feb 22 01:15:25 CET 2002
	- handle magic in sthaw.

0.1   Thu Sep 27 18:50:11 CEST 2001
	- compress/decompress considered stable.
        - added sfreeze, sfreeze_c and sfreeze_cr.

0.05  Thu Jan 25 12:47:36 CET 2001
	- uncompress leaked memory. doh!

0.04  Tue Jan 16 13:21:14 CET 2001
	- fix silly beginners 32/64 bit mistake.

0.03  Sat Jan 13 19:19:15 CET 2001
	- small testsuite fix (use Compress::LZF instead of LZV1!).

0.02  Thu Jan  4 05:51:26 CET 2001
	- upgraded to liblzf-0.2

0.01
	- improved version of Compress::LZV1, better compression, saner format
          freeer license, faster code...



lzf_c_best.c
/*
 * Copyright (c) 2000-2012 Marc Alexander Lehmann <schmorp@schmorp.de>
 *
 * Redistribution and use in source and binary forms, with or without modifica-
 * tion, are permitted provided that the following conditions are met:
 *
 *   1.  Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 *
 *   2.  Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MER-
 * CHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPE-
 * CIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTH-
 * ERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Alternatively, the contents of this file may be used under the terms of
 * the GNU General Public License ("GPL") version 2 or any later version,
 * in which case the provisions of the GPL are applicable instead of
 * the above. If you wish to allow the use of your version of this file
 * only under the terms of the GPL and not to allow others to use your
 * version of this file under the BSD license, indicate your decision
 * by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL. If you do not delete the
 * provisions above, a recipient may use your version of this file under
 * either the BSD or the GPL.
 */

#include "lzfP.h"

#define HASH(p) (p[0] << 6) ^ (p[1] << 3) ^ p[2]

#define        MAX_LIT        (1 <<  5)
#define        MAX_OFF        (1 << 13)
#define        MAX_REF        ((1 << 8) + (1 << 3))

#if __GNUC__ >= 3
# define expect(expr,value)         __builtin_expect ((expr),(value))
# define inline                     inline
#else
# define expect(expr,value)         (expr)
# define inline                     static
#endif

#define expect_false(expr) expect ((expr) != 0, 0)
#define expect_true(expr)  expect ((expr) != 0, 1)

/*
 * compressed format
 *
 * 000LLLLL <L+1>    ; literal, L+1=1..33 octets
 * LLLooooo oooooooo ; backref L+1=1..7 octets, o+1=1..4096 offset
 * 111ooooo LLLLLLLL oooooooo ; backref L+8 octets, o+1=1..4096 offset
 *
 */

unsigned int
lzf_compress_best (const void *const in_data, unsigned int in_len,
	           void *out_data, unsigned int out_len)
{
  const u8 *ip = (const u8 *)in_data;
        u8 *op = (u8 *)out_data;
  const u8 *in_end  = ip + in_len;
        u8 *out_end = op + out_len;

  const u8 *first [1 << (6+8)]; /* most recent occurance of a match */
  u16 prev [MAX_OFF]; /* how many bytes to go backwards for te next match */

  int lit;

  if (!in_len || !out_len)
    return 0;

  lit = 0; op++; /* start run */

  lit++; *op++ = *ip++;

  while (ip < in_end - 2)
    {
      int best_l = 0;
      const u8 *best_p;
      int e = (in_end - ip < MAX_REF ? in_end - ip : MAX_REF) - 1;
      unsigned int res = ((unsigned int)ip) & (MAX_OFF - 1);
      u16 hash = HASH (ip);
      u16 diff;
      const u8 *b = ip < (u8 *)in_data + MAX_OFF ? in_data : ip - MAX_OFF;
      const u8 *p = first [hash];
      prev [res] = ip - p; /* update ptr to previous hash match */
      first [hash] = ip; /* first hash match is here */

      if (p < ip)
      while (p >= b)
        {
          if (p[2] == ip[2]) /* first two bytes almost always match */
            if (p[best_l] == ip[best_l]) /* new match must be longer than the old one to qualify */
              if (p[1] == ip[1] && p[0] == ip[0]) /* just to be sure */
                {
                  int l = 3;

                  while (p[l] == ip[l] && l < e)
                    ++l;

                  if (l >= best_l)
                    {
                      best_p = p;
                      best_l = l;
                    }
                }

          diff = prev [((unsigned int)p) & (MAX_OFF - 1)];
          p = diff ? p - diff : 0;
        }

      if (best_l)
        {
          int len = best_l;
          int off = ip - best_p - 1;

          if (expect_false (op + 3 + 1 >= out_end)) /* first a faster conservative test */
            if (op - !lit + 3 + 1 >= out_end) /* second the exact but rare test */
              return 0;

          op [- lit - 1] = lit - 1; /* stop run */
          op -= !lit; /* undo run if length is zero */

          len -= 2; /* len is now #octets - 1 */
          ip++;

          if (len < 7)
            {
              *op++ = (off >> 8) + (len << 5);
            }
          else
            {
              *op++ = (off >> 8) + (  7 << 5);
              *op++ = len - 7;
            }

          *op++ = off;

          lit = 0; op++; /* start run */

          ip += len + 1;

          if (expect_false (ip >= in_end - 2))
            break;

          ip -= len + 1;

          //printf (" fill %p for %d\n", ip, len);//D
          do
            {
              u16 hash = HASH (ip);
              res = ((unsigned int)ip) & (MAX_OFF - 1);

              p = first [hash];
              prev [res] = ip - p; /* update ptr to previous hash match */
              first [hash] = ip; /* first hash match is here */

              ip++;
            }
          while (len--);
        }
      else
        {
          /* one more literal byte we must copy */
          if (expect_false (op >= out_end))
            return 0;

          lit++; *op++ = *ip++;

          if (expect_false (lit == MAX_LIT))
            {
              op [- lit - 1] = lit - 1; /* stop run */
              lit = 0; op++; /* start run */
            }
        }
    }

  if (op + 3 > out_end) /* at most 3 bytes can be missing here */
    return 0;

  while (ip < in_end)
    {
      lit++; *op++ = *ip++;

      if (expect_false (lit == MAX_LIT))
        {
          op [- lit - 1] = lit - 1; /* stop run */
          lit = 0; op++; /* start run */
        }
    }

  op [- lit - 1] = lit - 1; /* end run */
  op -= !lit; /* undo run if length is zero */

  return op - (u8 *)out_data;
}



LZF.pm
=head1 NAME

Compress::LZF - extremely light-weight Lempel-Ziv-Free compression

=head1 SYNOPSIS

   # import compress/decompress functions
   use Compress::LZF;
   # the same as above
   use Compress::LZF ':compress';

   $compressed = compress $uncompressed_data;
   $original_data = decompress $compressed;

   # import sfreeze, sfreeze_cref and sfreeze_c
   use Compress::LZF ':freeze';

   $serialized = sfreeze_c [4,5,6];
   $original_data = sthaw $serialized;

=head1 DESCRIPTION

LZF is an extremely fast (not that much slower than a pure memcpy)
compression algorithm. It is ideal for applications where you want to save
I<some> space but not at the cost of speed. It is ideal for repetitive
data as well. The module is self-contained and very small (no large
library to be pulled in). It is also free, so there should be no problems
incorporating this module into commercial programs.

I have no idea wether any patents in any countries apply to this
algorithm, but at the moment it is believed that it is free from any
patents.

=head1 FUNCTIONS

=over 4

=item $compressed = compress $uncompressed

=item $compressed = compress_best $uncompressed

Try to compress the given string as quickly and as much as possible. In
the worst case, the string can enlarge by 1 byte, but that should be the
absolute exception. You can expect a 45% compression ratio on large,
binary strings.

The C<compress_best> function uses a different algorithm that is slower
but usually achieves better compression.

=item $decompressed = decompress $compressed

Uncompress the string (compressed by C<compress>) and return the original
data. Decompression errors can result in either broken data (there is no
checksum kept) or a runtime error.

=item $serialized = sfreeze $value (simplified freeze)

=item $serialized = sfreeze_best $value

Often there is the need to serialize data into a string. This function does that, by using the Storable
module. It does the following transforms:

  undef (the perl undefined value)
     => a special cookie (undef'ness is being preserved)
  IV, NV, PV (i.e. a _plain_ perl scalar):
     => stays as is when it contains normal text/numbers
     => gets serialized into a string
  RV, undef, other funny objects (magical ones for example):
     => data structure is freeze'd into a string.

That is, it tries to leave "normal", human-readable data untouched but
still serializes complex data structures into strings. The idea is to keep
readability as high as possible, and in cases readability can't be helped
anyways, it tries to compress the string.

The C<sfreeze> functions will enlarge the original data one byte at most
and will only load the Storable method when neccessary.

The C<sfreeze_best> function uses a different algorithm that is slower
but usually achieves better compression.

=item $serialized = sfreeze_c $value (sfreeze and compress)

=item $serialized = sfreeze_c_best $value

Similar to C<sfreeze>, but always tries to C<c>ompress the resulting
string. This still leaves most small objects (most numbers) untouched.

The C<sfreeze_c> function uses a different algorithm that is slower
but usually achieves better compression.

=item $serialized = sfreeze_cr $value (sfreeze and compress references)

=item $serialized = sfreeze_cr_best $value

Similar to C<sfreeze>, but tries to C<c>ompress the resulting string
unless it's a "simple" string. References for example are not "simple" and
as such are being compressed.

The C<sfreeze_cr_best> function uses a different algorithm that is slower
but usually achieves better compression.

=item $original_data = sthaw $serialized

Recreate the original object from it's serialized representation. This
function automatically detects all the different sfreeze formats.

=item Compress::LZF::set_serializer $package, $freeze, $thaw

Set the serialize module and functions to use. The default is "Storable",
"Storable::net_mstore" and "Storable::mretrieve", which should be fine for
most purposes.

=back

=head1 SEE ALSO

Other Compress::* modules, especially Compress::LZV1 (an older, less
speedy module that guarentees only 1 byte overhead worst case) and
Compress::Zlib.

http://liblzf.plan9.de/

=head1 AUTHOR

This perl extension and the underlying liblzf were written by Marc Lehmann
<schmorp@schmorp.de> (See also http://liblzf.plan9.de/).

=head1 BUGS

=cut

package Compress::LZF;

require Exporter;
require DynaLoader;

$VERSION = '3.7';
@ISA = qw/Exporter DynaLoader/;
%EXPORT_TAGS = (
      freeze   => [qw(sfreeze sfreeze_best sfreeze_cr sfreeze_cr_best sfreeze_c sfreeze_c_best sthaw)],
      compress => [qw(compress compress_best decompress)],
);

Exporter::export_tags('compress');
Exporter::export_ok_tags('freeze');

bootstrap Compress::LZF $VERSION;

1;





